matrix/tasks/health.py

```diff
@@ -19,22 +22,27 @@
             workload_status_duration = now - unit.workload_status_since
             workload_busy = workload_status_duration < stable_period
             unit_busy = agent_busy or workload_busy
+            agent_idle = unit.agent_status == 'idle'
+            workload_ready = unit.workload_status in ('active', 'unknown')
 
             if unit.workload_status == 'error':
                 errored_units.append(unit)
-            elif (unit_busy or
-                  unit.agent_status != 'idle' or
-                  unit.workload_status not in ('active', 'unknown')):
+            elif workload_ready and agent_idle and unit_busy:
+                settling_units.append(unit)
+            elif not workload_ready or not agent_idle or unit_busy:
                 busy_units.append(unit)
 
     if errored_apps or errored_units:
         result = 'unhealthy'
     elif (busy_apps or busy_units) and \
             context.states.get('health.status') != 'healthy':
         result = 'busy'
+    elif settling_units and \
+            context.states.get('health.status') != 'healthy':
+        result = 'settling'
     else:
         result = 'healthy'
 
     context.set_state('health.status', result)
     rule.log.info("Health check: %s", result)
-    return True
+    return result != 'unhealthy'
```

Comment:
I think True here is the right answer. Did this plugin exist successfully is the question. The semantics of the return are not about the work the task did, only if it could or could not. 

There might be many occasions where the task is functioning properly but the result isn't as expected.  For the type of signaling you expect here I think you'd set a state.